warehouse/forklift/legacy.py

```diff
@@ -678,17 +680,55 @@ def _no_deprecated_classifiers(request):
         )
     }
 
+    def get_alternatives(classifier):
+        queue = collections.deque()
+        queue.append(classifier)
+
+        found_alternatives = []
+        visited_classifiers = {classifier.id}
+
+        while queue:
+            classifier = queue.popleft()
+
+            for alternative_classifier in classifier.alternatives:
+                if alternative_classifier.id in visited_classifiers:
+                    continue
+                visited_classifiers.add(alternative_classifier.id)
+                if alternative_classifier.deprecated:
+                    queue.append(alternative_classifier)
+                else:
+                    found_alternatives.append(alternative_classifier)
+
+        return found_alternatives
+
     def validate_no_deprecated_classifiers(form, field):
         invalid_classifiers = set(field.data or []) & deprecated_classifiers
         if invalid_classifiers:
-            first_invalid_classifier = sorted(invalid_classifiers)[0]
+            first_invalid_classifier = min(invalid_classifiers)
             host = request.registry.settings.get("warehouse.domain")
             classifiers_url = request.route_url("classifiers", _host=host)
+            alternatives = get_alternatives(
+                request.db.query(Classifier)
+                .filter(Classifier.classifier == first_invalid_classifier)
+                .limit(1)
+                .one()
+            )
+            alternatives.sort(key=operator.attrgetter("classifier"))
+
+            alternative_classifiers = ""
+            if alternatives:
+                alternative_classifiers = (
+                    f", and replaced with the following "
+                    f"classifier{'s' if len(alternatives) > 1 else ''}: "
+                    + ", ".join(
+                        f"{classifier.classifier!r}" for classifier in alternatives
+                    )
+                )
 
             raise wtforms.validators.ValidationError(
                 f"Classifier {first_invalid_classifier!r} has been "
-                f"deprecated, see {classifiers_url} for a list of valid "
-                "classifiers."
+                f"deprecated{alternative_classifiers}. "
+                f"See {classifiers_url} for a list of valid classifiers."
```

Comment:
I think this validation message should say:

> Classifier "Whatever" has been deprecated, see https://pypi.org/classifiers/ for a list of valid classifiers.

when there is no alternative, or

> Classifier "Whatever" has been deprecated in favor of the following classifier(s): "Something", "Another".

when there are alternatives. The logic can just be simplified to:

```python
if alternatives:
    raise ...
else:
    raise ...
```